crates/sui-core/src/transaction_orchestrator.rs

```diff
@@ -259,6 +243,15 @@ where
             .await?
         {
             debug!(?tx_digest, "no pending request in flight, submitting.");
+            let _txn_finality_timer = if transaction.contains_shared_object() {
```

Comment:
this function returns an awaitable, so when it ends, it does not guarantee the tx finishes. I suggest we keep `fn get_timer_guards` but instead return a tuple of `HistogramTimerGuard`, then explicitly drops the `wait_for_finality` one before `match result`